src/main/java/org/opensearch/security/configuration/ConfigurationRepository.java

```diff
@@ -119,8 +141,33 @@ private ConfigurationRepository(
         this.configurationChangedListener = new ArrayList<>();
         this.acceptInvalid = settings.getAsBoolean(ConfigConstants.SECURITY_UNSUPPORTED_ACCEPT_INVALID_CONFIG, false);
         cl = new ConfigurationLoaderSecurity7(client, threadPool, settings, clusterService);
-
         configCache = CacheBuilder.newBuilder().build();
+        this.securityIndexHandler = securityIndexHandler;
+    }
+
+    private Path resolveConfigDir() {
+        return Optional.ofNullable(System.getProperty("security.default_init.dir"))
+            .map(Path::of)
+            .orElseGet(() -> new Environment(settings, configPath).configDir().resolve("opensearch-security/"));
+    }
+
+    @Override
+    public void clusterChanged(final ClusterChangedEvent event) {
+        final SecurityMetadata securityMetadata = event.state().custom(SecurityMetadata.TYPE);
+        if (nodeSelectedAsManager(event)) {
+            if (securityMetadata == null) {
+                initSecurityIndex(event);
+            }
+        }
+        if (securityMetadata != null) {
+            executeConfigurationInitialization(securityMetadata);
```

Comment:
This looks much simpler. Nice!

Just to make sure I'm understanding this correctly, one of the primary purposes of `executeConfigurationInitialization` is to notify all of the nodes to populate their caches? 

On first bootstrap, `initSecurityIndex` is called since securityMetadata is null and it triggers [clusterService.submitStateUpdateTask](https://github.com/opensearch-project/security/pull/4002/files#diff-3dfaf08b5fd593ac73bb4955f7f579919ac870b5ed060636a2320e3ef85ef935R379-R387) which will again hit this block, but the second time it hits this block the securityMetadata will be non-null so it triggers this call?

Reply:
Exactly. As soon as cluster has a manager. The manager node triggers index initialization and if all is good it changes the cluster state so all nodes including manages will reload the cache. The cluster state update needs to be improved a bit but it wont change the current logic we have right now.

Reply:
done